consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,256 @@
+	r := interactor
+	if interactor == nil {
+		r = &defaultRootchainInteractor{}
+	}
+
+	return &checkpointManager{
+		sender:            sender,
+		blockchain:        blockchain,
+		consensusBackend:  backend,
+		rootchain:         r,
+		checkpointsOffset: uint64(checkpointTimeInterval.Milliseconds() / blockTime.Milliseconds()),
+	}
+}
+
+// getCurrentCheckpointID queries CheckpointManager smart contract and retrieves current checkpoint id
+func (c checkpointManager) getCurrentCheckpointID(epochNumber uint64) (uint64, error) {
+	checkpointIDMethodEncoded, err := currentCheckpointIDMethod.Encode([]interface{}{})
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number for epoch=%d: %w",
+			currentCheckpointID, epochNumber, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, epochNumber uint64) error {
+	checkpointID, err := c.getCurrentCheckpointID(epochNumber)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNum := checkpointID + 1; blockNum < latestHeader.Number; blockNum++ {
+		header, exists := c.blockchain.GetHeaderByNumber(blockNum)
+		if !exists {
+			return fmt.Errorf("block %d was not found", blockNum)
+		}
+
+		isEndOfEpoch, extra, err := c.isEndOfEpoch(*header)
+		if err != nil {
+			return err
+		}
+		// send pending checkpoints only for epoch ending blocks
+		if !isEndOfEpoch {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *header, *extra)
+		if err != nil {
+			return err
+		}
+		nonce++
+	}
+
+	extra, err := GetIbftExtra(latestHeader.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	return c.submitCheckpointInternal(nonce, txn, latestHeader, *extra)
+}
+
+// submitCheckpointInternal encodes checkpoint data for the given block and
+// sends a transaction to the CheckpointManager rootchain contract
+func (c *checkpointManager) submitCheckpointInternal(nonce uint64, txn *ethgo.Transaction,
+	header types.Header, extra Extra) error {
+	nextEpochValidators := AccountSet{}
+
+	isEndOfEpoch, _, err := c.isEndOfEpoch(header)
```

Comment:
This check will be changed as well if you are to take in consideration this [comment](https://github.com/0xPolygon/polygon-edge/pull/861/files#r1012748750).

Reply:
Fixed [bc66984](https://github.com/0xPolygon/polygon-edge/pull/861/commits/bc66984233759009527557df6d0951b5fdc6239c)